File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/checkin/map/MapViewModel.kt
package com.hhp227.concafe.presentation.main.checkin.map

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.CafeDetailEvent
import com.hhp227.concafe.domain.event.publisher.CafeDetailEventPublisher
import com.hhp227.concafe.domain.model.Cafe
import com.hhp227.concafe.domain.usecase.GetCheckInGuestFeedUseCase
import com.hhp227.concafe.presentation.main.checkin.CheckInLocationProvider
import com.hhp227.concafe.presentation.main.checkin.CheckInLocationResult
import com.hhp227.concafe.presentation.main.explore.ExploreUiState
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class MapViewModel(
    private val getCheckInGuestFeedUseCase: GetCheckInGuestFeedUseCase,
    private val cafeDetailEventPublisher: CafeDetailEventPublisher,
    private val checkInLocationProvider: CheckInLocationProvider
) : ViewModel() {
    private val _uiState = MutableStateFlow(MapUiState.empty())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<MapEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun loadMapFeed() {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }

        jobs[TaskKey.LOAD_MAP_FEED]?.cancel()
        jobs[TaskKey.LOAD_MAP_FEED] = viewModelScope.launch {
            when (val result = getCheckInGuestFeedUseCase.invoke()) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = null,
                            currentLocationLabel = result.data.currentLocationLabel,
                            mapCafes = result.data.mapCafes
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    private fun detectUserCity() {
        jobs[TaskKey.DETECT_CITY]?.cancel()
        jobs[TaskKey.DETECT_CITY] = viewModelScope.launch {
            if (_uiState.value.userCityKey != null) return@launch

            when (val result = checkInLocationProvider.getCurrentLocation()) {
                is CheckInLocationResult.Success -> {
                    val cityKey = cityKeyFromCoordinates(
                        lat = result.location.latitude,
                        lng = result.location.longitude
                    )
                    _uiState.update { it.copy(userCityKey = cityKey) }
                }
                is CheckInLocationResult.Failure -> Unit
            }
        }
    }

    private fun observeCafeDetailEvent() {
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT] = viewModelScope.launch {
            cafeDetailEventPublisher.events.collectLatest { event ->
                if (event is CafeDetailEvent.CafeInfoUpdated) {
                    patchCafe(event.cafe)
                }
            }
        }
    }

    private fun patchCafe(cafe: Cafe) {
        _uiState.update { state ->
            state.copy(
                mapCafes = state.mapCafes.map { item ->
                    if (item.id == cafe.id) {
                        item.copy(name = cafe.name, locationLabel = cafe.region.city, rating = cafe.ratingAvg)
                    } else {
                        item
                    }
                }
            )
        }
    }

    fun initializeRegion(regionKey: String?) {
        val region = regionKey
            ?.let { key -> ExploreUiState.RegionFilter.entries.firstOrNull { it.name == key } }
            ?: return

        _uiState.update { state ->
            if (state.selectedRegion == ExploreUiState.RegionFilter.ALL) {
                state.copy(selectedRegion = region)
            } else {
                state
            }
        }
    }

    fun onAction(action: MapAction) {
        viewModelScope.launch {
            when (action) {
                MapAction.ClickBack -> _event.emit(MapEvent.NavigateBack)
                is MapAction.ClickCafe -> _event.emit(MapEvent.NavigateToCafe(action.id))
                is MapAction.UpdateRegion -> _uiState.update { it.copy(selectedRegion = action.region) }
                is MapAction.UpdateSearchQuery -> _uiState.update { it.copy(searchQuery = action.query) }
            }
        }
    }

    init {
        observeCafeDetailEvent()
        detectUserCity()
        loadMapFeed()
    }

    override fun onCleared() {
        jobs.values.forEach { it.cancel() }
        jobs.clear()
        super.onCleared()
    }

    private enum class TaskKey {
        LOAD_MAP_FEED,
        OBSERVE_CAFE_DETAIL_EVENT,
        DETECT_CITY
    }

    private fun cityKeyFromCoordinates(lat: Double, lng: Double): String? {
        return when {
            lat in 37.4..37.7 && lng in 126.7..127.2 -> "seoul"
            lat in 35.0..35.4 && lng in 128.8..129.3 -> "busan"
            lat in 35.7..36.0 && lng in 128.4..128.8 -> "daegu"
            lat in 35.35..35.60 && lng in 139.50..139.75 -> "yokohama"
            lat in 35.5..35.9 && lng in 139.3..139.9 -> "tokyo"
            lat in 34.5..34.9 && lng in 135.3..135.7 -> "osaka"
            else -> null
        }
    }
}

File: iosApp/iosApp/Presentation/Main/CheckIn/Map/MapViewModel.swift
//
//  MapViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 4/23/26.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class MapViewModel: ObservableObject {
    private let getCheckInGuestFeedUseCase: GetCheckInGuestFeedUseCase

    private let getCheckInMapCafePageUseCase: GetCheckInMapCafePageUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let cafeDetailEventPublisher: CafeDetailEventPublisher

    private let currentLocationProvider = IosCheckInLocationProvider()

    @Published private(set) var uiState = MapUiState.empty

    let event = PassthroughSubject<MapEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func loadMapFeed() {
        uiState.isLoading = true
        uiState.errorMessage = nil

        tasks[.mapFeed]?.cancel()
        tasks[.mapFeed] = Task {
            do {
                let result = try await getCheckInGuestFeedUseCase.invoke()

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? Shared.CheckInGuestFeed {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                    uiState.currentLocationLabel = feed.currentLocationLabel
                    if uiState.selectedRegion == .all && uiState.userCityKey == nil {
                        uiState.mapCafes = feed.mapCafes
                    }
                } else if let failure = result as? AppResultFailure {
                    uiState.isLoading = false
                    uiState.errorMessage = "\(failure.error)"
                } else {
                    uiState.isLoading = false
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = error.localizedDescription
            }
        }
    }

    private func observeSession() {
        tasks[.session]?.cancel()
        tasks[.session] = Task {
            do {
                for try await user in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    uiState.isLoggedIn = user != nil
                    if user != nil {
                        uiState.isLoginPromptVisible = false
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func detectUserCity() {
        guard uiState.userCityKey == nil else { return }
        tasks[.detectCity]?.cancel()
        tasks[.detectCity] = Task {
            if let cached = currentLocationProvider.getLastKnownLocation() {
                let cityKey = Self.cityKeyFromCoordinates(
                    lat: cached.location.latitude,
                    lng: cached.location.longitude
                )
                if cityKey != nil {
                    uiState.userCityKey = cityKey
                    if uiState.selectedRegion == .all,
                       let cityKey {
                        loadMapCafesForRegion(cityKey)
                    }
                    return
                }
            }

            let result = await currentLocationProvider.getCurrentLocation()
            guard result.isSuccess else { return }
            let cityKey = Self.cityKeyFromCoordinates(
                lat: result.location.latitude,
                lng: result.location.longitude
            )
            uiState.userCityKey = cityKey
            if uiState.selectedRegion == .all,
               let cityKey {
                loadMapCafesForRegion(cityKey)
            }
        }
    }

    private func loadMapCafesForRegion(_ regionKey: String) {
        tasks[.mapRegion]?.cancel()
        tasks[.mapRegion] = Task {
            do {
                let result = try await getCheckInMapCafePageUseCase.invoke(regionKey: regionKey, pageSize: 80)
                if let success = result as? AppResultSuccess<AnyObject>,
                   let cafes = success.data as? [CheckInCafeSummary] {
                    uiState.mapCafes = cafes
                    uiState.errorMessage = nil
                } else if let failure = result as? AppResultFailure {
                    uiState.errorMessage = "\(failure.error)"
                }
            } catch {
                if Task.isCancelled { return }
                uiState.errorMessage = error.localizedDescription
            }
        }
    }

    private func observeCafeDetailEvent() {
        tasks[.cafeDetailEvent]?.cancel()
        tasks[.cafeDetailEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeDetailEventPublisher.events) {
                    if let updated = event as? CafeDetailEvent.CafeInfoUpdated {
                        patchCafe(updated.cafe)
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func patchCafe(_ cafe: Cafe) {
        uiState.mapCafes = uiState.mapCafes.map { item in
            guard item.id == cafe.id else { return item }
            return CheckInCafeSummary(
                id: item.id,
                name: cafe.name,
                locationLabel: cafe.region.city,
                geoPoint: item.geoPoint,
                rating: cafe.ratingAvg,
                checkInCount: item.checkInCount,
                thumbnailImage: item.thumbnailImage
            )
        }
    }

    func initializeRegion(_ regionKey: String?) {
        guard uiState.selectedRegion == .all,
              let regionKey,
              let region = ExploreUiState.RegionFilter.allCases.first(where: { $0.label == regionKey || $0.rawValue == regionKey || (regionKey == "yokohama" && $0 == .etc) }) else { return }
        uiState.selectedRegion = region
        loadMapCafesForRegion(region.rawValue)
    }

    func onAction(_ action: MapAction) {
        switch action {
        case .backTapped:
            event.send(.navigateBack)
        case .cafeTapped(let id):
            if uiState.isLoggedIn {
                event.send(.navigateToCafe(id: id))
            } else {
                uiState.isLoginPromptVisible = true
            }
        case .loginPromptSignInTapped:
            uiState.isLoginPromptVisible = false
            event.send(.navigateToSignIn)
        case .dismissLoginPrompt:
            uiState.isLoginPromptVisible = false
        case .regionChanged(let region):
            uiState.selectedRegion = region
            if region == .all {
                if let userCityKey = uiState.userCityKey {
                    loadMapCafesForRegion(userCityKey)
                }
            } else {
                loadMapCafesForRegion(region.rawValue)
            }
        case .searchQueryChanged(let query):
            uiState.searchQuery = query
        }
    }

    init(
        getCheckInGuestFeedUseCase: GetCheckInGuestFeedUseCase = KoinInitializerKt.resolveGetCheckInGuestFeedUseCase(),
        getCheckInMapCafePageUseCase: GetCheckInMapCafePageUseCase = KoinInitializerKt.resolveGetCheckInMapCafePageUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        cafeDetailEventPublisher: CafeDetailEventPublisher = KoinInitializerKt.resolveCafeDetailEventPublisher()
    ) {
        self.getCheckInGuestFeedUseCase = getCheckInGuestFeedUseCase
        self.getCheckInMapCafePageUseCase = getCheckInMapCafePageUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.cafeDetailEventPublisher = cafeDetailEventPublisher

        observeSession()
        observeCafeDetailEvent()
        detectUserCity()
        loadMapFeed()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case session
        case mapFeed
        case mapRegion
        case cafeDetailEvent
        case detectCity
    }

    private static func cityKeyFromCoordinates(lat: Double, lng: Double) -> String? {
        if (37.4...37.7).contains(lat) && (126.7...127.2).contains(lng) { return "seoul" }
        if (35.0...35.4).contains(lat) && (128.8...129.3).contains(lng) { return "busan" }
        if (35.7...36.0).contains(lat) && (128.4...128.8).contains(lng) { return "daegu" }
        if (35.35...35.60).contains(lat) && (139.50...139.75).contains(lng) { return "etc" }
        if (35.5...35.9).contains(lat) && (139.3...139.9).contains(lng) { return "tokyo" }
        if (34.5...34.9).contains(lat) && (135.3...135.7).contains(lng) { return "osaka" }
        return nil
    }
}
